packages/@aws-cdk/core/lib/private/refs.ts

```diff
@@ -0,0 +1,276 @@
+  }
+}
+
+/**
+ * Resolves the value for `reference` in the context of `consumer`.
+ */
+function resolveValue(consumer: Stack, reference: CfnReference): IResolvable {
+  const producer = Stack.of(reference.target);
+
+  // produce and consumer stacks are the same, we can just return the value itself.
+  if (producer === consumer) {
+    return reference;
+  }
+
+  // unsupported: stacks from different apps
+  if (producer.node.root !== consumer.node.root) {
+    throw new Error('Cannot reference across apps. Consuming and producing stacks must be defined within the same CDK app.');
+  }
+
+  // unsupported: stacks are not in the same environment
+  if (producer.environment !== consumer.environment) {
+    throw new Error(
+      `Stack "${consumer.node.path}" cannot consume a cross reference from stack "${producer.node.path}". ` +
+      'Cross stack references are only supported for stacks deployed to the same environment or between nested stacks and their parent stack');
+  }
+
+  // ----------------------------------------------------------------------
+  // consumer is nested in the producer (directly or indirectly)
+  // ----------------------------------------------------------------------
+
+  // if the consumer is a child of the producer, wire the reference through a
+  // CloudFormation parameter on the consumer and resolve recursively.
+  if (isParent(producer, consumer)) {
+    const parameterValue = createNestedStackParameter(consumer, reference);
+    return resolveValue(consumer, parameterValue);
+  }
+
+  // ----------------------------------------------------------------------
+  // producer is a nested stack
+  // ----------------------------------------------------------------------
+
+  // if the producer is nested, always publish the value through a
+  // cloudformation output and resolve recursively with the Fn::GetAtt
+  // of the output in the parent stack.
+
+  // one might ask, if the consumer is not a parent of the producer,
+  // why not just use export/import? the reason is that we cannot
+  // generate an "export name" from a nested stack because the export
+  // name must contain the stack name to ensure uniqueness, and we
+  // don't know the stack name of a nested stack before we deploy it.
+  // therefore, we can only export from a top-level stack.
+  if (producer.nested) {
+    const outputValue = createNestedStackOutput(producer, reference);
+    return resolveValue(consumer, outputValue);
+  }
+
+  // ----------------------------------------------------------------------
+  // export/import
+  // ----------------------------------------------------------------------
+
+  // export the value through a cloudformation "export name" and use an
+  // Fn::ImportValue in the consumption site.
+
+  // delcare a dependency between the two top-level (non-nested) stacks to make
+  // sure the producer is deployed before the consumer.
+  const producerDep = producer.nestedStackParent ?? producer;
```

Comment:
Bug? What is `.nestedStackParent` is a nested stack? We need to find the top level!

Reply:
Will `producer.nestedStackParent` ever be `true`? 
In line 87 we handle `producer.nested` case

Reply:
If you meant the consumer, then yes, I think there is a bug here. 
The only thing we know about the consumer is that it's not a child of the producer, but it can still be a nested stack which means we need to bubble the parameter up.

Reply:
This is not about the parameter, it’s about the dependency, but yes. 

I actually want to move the dependency resolution logic to `app.prepare` loop as well, so I can simply specify the dependencies between the two nodes and the dep resolution logic will take care of the rest. But I will do it in a subsequent PR once this is merged.

Reply:
I think this will happen on the consumer though. All we know on the consumer is that **it is not a child of the producer**, but it still can be a nested stack which means we need to create the dependency on its root parent.

Reply:
Yes, you are right. At any rate I changed this to:

```ts
const producerDep = findTopLevel(producer);
const consumerDep = findTopLevel(consumer);
```

So we always define the dependency between the top-level stacks.

Reply:
Okay this can be even simpler because `stack.addDependency` has all the logic to handle nested stacks already, so this became `consumer.addDependency(producer)`. yey!